File: Day14/Part2/bfs_dynamic.cpp
#include <bits/stdc++.h>
using namespace std;

// Part 2 — BFS Traversal with Dynamic Portals (deterministic remap per phase)

using Pos = pair<int,int>;

static vector<Pos> collectTargets(const vector<vector<char>>& maze, char target){
    vector<Pos> res;
    for (int i = 0; i < (int)maze.size(); ++i)
        for (int j = 0; j < (int)maze[i].size(); ++j)
            if (maze[i][j] == target) res.emplace_back(i,j);
    return res;
}

static vector<Pos> collectEmpties(const vector<vector<char>>& maze){
    vector<Pos> res;
    for (int i = 0; i < (int)maze.size(); ++i)
        for (int j = 0; j < (int)maze[i].size(); ++j)
            if (maze[i][j] == 'E' || maze[i][j] == 'P') res.emplace_back(i,j);
    return res;
}

static set<Pos> portalsForPhase(const vector<Pos>& empties, int pcount, int phase){
    set<Pos> s;
    if (pcount <= 0 || empties.empty()) return s;
    int m = (int)empties.size();
    for (int k = 0; k < pcount; ++k){
        s.insert(empties[(phase + k) % m]);
    }
    return s;
}

static optional<Pos> teleportDestination(const vector<Pos>& portals, Pos current){
    if (portals.size() <= 1) return nullopt;
    for (size_t i = 0; i < portals.size(); ++i){
        if (portals[i] == current){
            size_t nxt = (i + 1) % portals.size();
            if (portals[nxt] == current) return nullopt;
            return portals[nxt];
        }
    }
    return nullopt;
}

int shortestPathWithPortals(vector<vector<char>>& maze, int t){
    int n = (int)maze.size();
    auto starts = collectTargets(maze, 'S');
    auto exits = collectTargets(maze, 'X');
    auto basePortals = collectTargets(maze, 'P');
    auto empties = collectEmpties(maze);
    if (starts.empty() || exits.empty()) return -1;
    Pos S = starts[0];
    Pos X = exits[0];
    int pcount = (int)basePortals.size();

    deque<tuple<int,int,int>> q; // x,y,steps
    q.emplace_back(S.first, S.second, 0);
    vector<vector<char>> vis(n, vector<char>(n, 0));
    vis[S.first][S.second] = 1;

    const int dx[4] = {1,-1,0,0};
    const int dy[4] = {0,0,1,-1};

    while(!q.empty()){
        auto [x,y,steps] = q.front(); q.pop_front();
        if (x == X.first && y == X.second) return steps;

        int phase = steps / t;
        auto pset = portalsForPhase(empties, pcount, phase);
        vector<Pos> plist(pset.begin(), pset.end());

        for (int d=0; d<4; ++d){
            int nx = x + dx[d], ny = y + dy[d];
            if (nx<0 || ny<0 || nx>=n || ny>=n) continue;
            char cell = maze[nx][ny];
            if (cell == 'W') continue;

            int tx = nx, ty = ny;
            if (pset.count({nx,ny})){
                auto dest = teleportDestination(plist, {nx,ny});
                if (dest) { tx = dest->first; ty = dest->second; }
            }

            if (!vis[tx][ty]){
                vis[tx][ty] = 1;
                q.emplace_back(tx, ty, steps+1);
            }
        }
    }
    return -1;
}

int main(){
    vector<vector<char>> maze = {
        {'S','E','P','W'},
        {'E','W','E','E'},
        {'E','P','W','E'},
        {'E','E','E','X'}
    };
    cout << shortestPathWithPortals(maze, 3) << "\n";
    return 0;
}

File: Day14/Part2/bfs_dynamic.py
from collections import deque
from typing import List, Tuple, Optional, Set

# Part 2 — BFS Traversal with Dynamic Portals
# Deterministic simulation of portal remapping every t steps.

Pos = Tuple[int, int]
Grid = List[List[str]]


def _collect_cells(maze: Grid):
    n = len(maze)
    starts: List[Pos] = []
    exits: List[Pos] = []
    portals: List[Pos] = []
    empties: List[Pos] = []
    for i in range(n):
        for j in range(len(maze[i])):
            c = maze[i][j]
            if c == 'S':
                starts.append((i, j))
            elif c == 'X':
                exits.append((i, j))
            elif c == 'P':
                portals.append((i, j))
            if c == 'E' or c == 'P':
                empties.append((i, j))
    return starts, exits, portals, empties


def _portals_for_phase(empties: List[Pos], pcount: int, phase: int) -> Set[Pos]:
    """Deterministic portal layout for a given phase index.
    We pick pcount cells from empties by a cyclic shift starting offset=phase.
    """
    if pcount <= 0 or not empties:
        return set()
    res = set()
    m = len(empties)
    for k in range(pcount):
        res.add(empties[(phase + k) % m])
    return res


def _teleport_destination(portals: List[Pos], current: Pos) -> Optional[Pos]:
    """Teleport to another portal deterministically: the next portal in list order.
    If only one portal exists or none, return None (no-op).
    If current is not in portals (due to mapping mismatch), treat as no-op.
    """
    if len(portals) <= 1:
        return None
    try:
        idx = portals.index(current)
    except ValueError:
        return None
    nxt = (idx + 1) % len(portals)
    if portals[nxt] == current:
        return None
    return portals[nxt]


def shortestPathWithPortals(maze: Grid, t: int) -> int:
    n = len(maze)
    starts, exits, base_portals, empties = _collect_cells(maze)
    if not starts or not exits:
        return -1
    S = starts[0]
    X = exits[0]
    pcount = len(base_portals)

    # BFS with simple 2D visited (Part 2 baseline). Portals remap by phase = steps // t.
    q = deque()
    q.append((S[0], S[1], 0))  # (x, y, steps)
    visited = [[False]*n for _ in range(n)]
    visited[S[0]][S[1]] = True

    dirs = [(1,0),(-1,0),(0,1),(0,-1)]

    while q:
        x, y, steps = q.popleft()
        if (x, y) == X:
            return steps

        phase = steps // t  # portals layout for this step window
        portals_set = _portals_for_phase(empties, pcount, phase)
        portals_list = sorted(list(portals_set))  # stable order

        for dx, dy in dirs:
            nx, ny = x + dx, y + dy
            if not (0 <= nx < n and 0 <= ny < n):
                continue
            cell = maze[nx][ny]
            if cell == 'W':
                continue

            tx, ty = nx, ny
            # If destination is a portal under current layout, teleport instantly (no extra step)
            if (nx, ny) in portals_set:
                dest = _teleport_destination(portals_list, (nx, ny))
                if dest is not None:
                    tx, ty = dest

            if not visited[tx][ty]:
                visited[tx][ty] = True
                q.append((tx, ty, steps + 1))

    return -1


if __name__ == "__main__":
    maze = [
        ['S','E','P','W'],
        ['E','W','E','E'],
        ['E','P','W','E'],
        ['E','E','E','X']
    ]
    print(shortestPathWithPortals(maze, 3))
